### PR TITLE
Added skew configuration for single thread kill timeout.

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

Commit Message: Added random skew option to watchdog kill_timeout.
Additional Description: 
Risk Level: Low, backwards compatible by default
Testing: Implemented unit tests
Docs Changes: No
Release Notes: None (this is fully backward compatible by default.)
Optional Fixes: #11389 (Fixes part of that issue)

### CODE REVIEW

**Comment by KBaichoo** [comment]
/ review antoniovicente

File: test/server/configuration_impl_test.cc [hunk 1]
```diff
@@ -730,6 +730,58 @@ TEST_F(ConfigurationImplTest, ExceedLoadBalancerLocalityWeightsLimit) {
...
+    },
+  })EOF";
+
+  envoy::config::bootstrap::v3::Bootstrap bootstrap;
+  TestUtility::loadFromJson(json, bootstrap);
+
+  MainImpl config;
+  config.initialize(bootstrap, server_, cluster_manager_factory_);
+
+  EXPECT_LT(std::chrono::milliseconds(1000), config.wdKillTimeout());
```
Comment (antoniovicente) [inline test/server/configuration_impl_test.cc:765]: Also:
EXPECT_GE(std::chrono::milliseconds(1500), config.wdKillTimeout());

File: source/server/configuration_impl.cc [hunk 2]
```diff
@@ -90,8 +90,21 @@ void MainImpl::initialize(const envoy::config::bootstrap::v3::Bootstrap& bootstr
...
-  watchdog_kill_timeout_ =
-      std::chrono::milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(watchdog, kill_timeout, 0));
+
+  uint64_t kill_timeout = PROTOBUF_GET_MS_OR_DEFAULT(watchdog, kill_timeout, 0);
+  const uint64_t max_skew = PROTOBUF_GET_MS_OR_DEFAULT(watchdog, max_kill_skew, 0);
+
+  // Adjust kill timeout if we have skew enabled.
+  if (kill_timeout > 0 && max_skew > 0) {
+
+    // Increments the kill timeout with a random value in (0, max_skew).
```
Comment (antoniovicente) [inline source/server/configuration_impl.cc:100]: The skew is in the closed interval [1 ms, max_skew in msec] or semi closed interval (0 ms, max_skew in msec].

I recommend changing ) to ]

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 3]
```diff
@@ -314,6 +315,10 @@ message Watchdog {
   // duration assume a true deadlock and kill the entire Envoy process. Set to 0
   // to disable this behavior. If not specified the default is 0 (disabled).
   google.protobuf.Duration multikill_timeout = 4;
+
+  // Defines the maximum skew used to adjust the kill_timeout if kill_timeout is
+  // enabled. Set to 0 to disable. If not specified the default is 0 (disabled).
```
Comment (antoniovicente) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:320]: Could you expand this comment with an explanation as to why you would configure this random additional watchdog single-thread kill timeout?  The reason is related to wanting to reduce risk of synchronized watchdog kill events across proxies due to external triggers.

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 4]
```diff
@@ -314,6 +315,10 @@ message Watchdog {
...
+  google.protobuf.Duration max_kill_skew = 5;
```
Comment (antoniovicente) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:321]: @envoyproxy/api-shepherds for naming suggestions and API approval.

Regarding parameter naming: In https://github.com/envoyproxy/envoy/issues/11389#issuecomment-657796431 you mention max_random_kill_duration which although not perfect seems more clear to me than the current parameter name.  I would have considered something along the lines of: random_additional_delay_before_kill_timeout

Comment (htuch) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:321]: I would be consistent with other places we add skew, e.g. `initial_jitter` in the health checkers. So maybe call this `max_kill_timeout_jitter`?

File: test/server/configuration_impl_test.cc (hunk 1 above)
Comment (KBaichoo) [reply test/server/configuration_impl_test.cc:765]: Done.

File: source/server/configuration_impl.cc (hunk 2 above)
Comment (KBaichoo) [reply source/server/configuration_impl.cc]: Good Catch. Done.

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 3 above)
Comment (KBaichoo) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Done.

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 4 above)
Comment (KBaichoo) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Will go with `max_kill_timeout_jitter` then for consistency.

**Comment by KBaichoo** [comment]
ping @antoniovicente

File: source/server/configuration_impl.cc [hunk 5]
```diff
@@ -90,8 +90,22 @@ void MainImpl::initialize(const envoy::config::bootstrap::v3::Bootstrap& bootstr
...
-  watchdog_kill_timeout_ =
-      std::chrono::milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(watchdog, kill_timeout, 0));
+
+  uint64_t kill_timeout = PROTOBUF_GET_MS_OR_DEFAULT(watchdog, kill_timeout, 0);
+  const uint64_t max_kill_timeout_jitter =
+      PROTOBUF_GET_MS_OR_DEFAULT(watchdog, max_kill_timeout_jitter, 0);
+
+  // Adjust kill timeout if we have skew enabled.
+  if (kill_timeout > 0 && max_kill_timeout_jitter > 0) {
+
```
Comment (antoniovicente) [inline source/server/configuration_impl.cc:100]: nit: I think we prefer to avoid newlines at the beginning of code blocks.

Comment (KBaichoo) [reply source/server/configuration_impl.cc]: Done.

**Comment by KBaichoo** [comment]
PTAL @envoyproxy/api-shepherds.

**Comment by KBaichoo** [comment]
/ re-test

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 6]
```diff
@@ -314,6 +315,12 @@ message Watchdog {
   // duration assume a true deadlock and kill the entire Envoy process. Set to 0
   // to disable this behavior. If not specified the default is 0 (disabled).
   google.protobuf.Duration multikill_timeout = 4;
+
+  // Defines the maximum skew used to adjust the kill_timeout if kill_timeout is
```
Comment (htuch) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:319]: Two nits: 
1. Can you move this next to `kill_timeout` for readability? It's fine to have discontinuous numbering.
2. For field names, either write `*kill_timeout*` if the field is nearby, or use a `:ref:` links (grep for `"ref:.*envoy_api_field"` for examples.

Comment (KBaichoo) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Done. Thanks for the pointers.

**Comment by antoniovicente** [comment]
/assign @alyssawilk 

Hey Alyssa, could you find a maintainer who can review and merge this change?  Harvey is OOO for 2 weeks.  Thanks!

**Comment by alyssawilk** [comment]
At a high level it looks good, but  guidelines say we need a cross-company sign-off so I'm not really useful from a review perspective.  How about @snowp ?

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 7]
```diff
@@ -316,6 +316,12 @@ message Watchdog {
   // kill behavior. If not specified the default is 0 (disabled).
   google.protobuf.Duration kill_timeout = 3;
 
+  // Defines the maximum skew used to adjust the *kill_timeout* if *kill_timeout* is
+  // enabled. Enabling this feature would help to reduce risk of synchronized
+  // watchdog kill events across proxies due to external triggers. Set to 0 to
+  // disable. If not specified the default is 0 (disabled).
+  google.protobuf.Duration max_kill_timeout_jitter = 6;
```
Comment (snowp) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:323]: Maybe add a pgv validation to ensure that this is gte 0?

looks something like this
```
  google.protobuf.Duration session_timeout = 6 [(validate.rules).duration = {
    lt {seconds: 4294967296}
    gte {}
  }];
```

Comment (KBaichoo) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Good idea, done.

**Comment by KBaichoo** [comment]
Kicking CI again, as the test that was failing is know to be flaky (see https://github.com/envoyproxy/envoy/issues/11851)

**Comment by KBaichoo** [comment]
I was able to run `//test/integration:tcp_proxy_integration_test` which failed `FAILED in 1 out of 2 in 128.9s` in the pipeline, but when I ran it locally 500 times, it didn't fail. 

Seems like it was a fluke, what should I do next? @snowp

**Comment by snowp** [comment]
Probably a flake, I'll rerun it. 

This needs @envoyproxy/api-shepherds review after the changes

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 8]
```diff
@@ -323,6 +323,12 @@ message Watchdog {
   // kill behavior. If not specified the default is 0 (disabled).
   google.protobuf.Duration kill_timeout = 3;
 
+  // Defines the maximum skew used to adjust the *kill_timeout* if *kill_timeout* is
```
Comment (mattklein123) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:326]: nit (we generally call this jitter everywhere else and the field references jitter)

```suggestion
  // Defines the maximum jitter used to adjust the *kill_timeout* if *kill_timeout* is
```

Also needs a release note.

**Comment by KBaichoo** [comment]
Whoops, I accidentally had issues with the DCO and followed the process it suggested -- but think this will be a bigger headache. Will just move this to another PR.

**Pull request closed without merging**